### PR TITLE
[KEYCLOAK-8742] [core] Allow OpenShift client to recognize custom OpenShift route subdomain

### PR DESCRIPTION
**Pull request by iankko**
(`xtf.config.route_domain` parameter setting), when generating hostname for the client

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

### CODE REVIEW

File: core/src/main/java/cz/xtf/core/openshift/OpenShift.java [hunk 1]
```diff
@@ -407,6 +408,10 @@ private String retrieveRouteSuffix() {
 		route = createRoute(route);
 		deleteRoute(route);
 
+		if (StringUtils.isNotBlank(cz.xtf.core.config.OpenShiftConfig.clientRouteDomain())) {
```
[the remark this answers is not in the document]
Comment (Crumby) [reply core/src/main/java/cz/xtf/core/openshift/OpenShift.java]: `cz.xtf.core.config.OpenShiftConfig.clientRouteDomain() != null` should be enough, but I'm not sure about that.

[the remark this answers is not in the document]
Comment (iankko) [reply core/src/main/java/cz/xtf/core/openshift/OpenShift.java]: @Crumby Thanks, Pavel! The check moved. Regarding the second != null check, wouldn't it allow (succeed) also if empty string was provided? Kept the StringUtils check form there, just moved it.

File: core/src/main/java/cz/xtf/core/config/OpenShiftConfig.java [hunk 2]
```diff
@@ -10,6 +10,7 @@
 	public static final String OPENSHIFT_ADMIN_PASSWORD = "xtf.openshift.admin.password";
 	public static final String OPENSHIFT_MASTER_USERNAME = "xtf.openshift.master.username";
 	public static final String OPENSHIFT_MASTER_PASSWORD = "xtf.openshift.master.password";
+	public static final String OPENSHIFT_CLIENT_CUSTOM_ROUTE_DOMAIN = "xtf.config.route_domain";
```
[the remark this answers is not in the document]
Comment (iankko) [reply core/src/main/java/cz/xtf/core/config/OpenShiftConfig.java]: Renamed all of variable name, property name, and method name to `routeDomain()`. Tried to add the backward compatibility check (based on existing examples & approach how the `routeDomain()` method is implemented in `utilities/src/main/java/cz/xtf/XTFConfiguration.java` path). But unsure it's OK, please double-check.

**Comment by iankko** [comment]
Updated the PR, but kept the second commit separate (so you can better see the diffs). Can squash the 2nd one, if desired (let me know).

File: core/src/main/java/cz/xtf/core/config/BackwardCompatibility.java [hunk 3]
```diff
@@ -72,4 +73,12 @@ private static String adminPassword() {
 
 		return password1 != null ? password1 : password2;
 	}
+
+	private static String routeDomain() {
+		String configDomain = XTFConfig.get("xtf.config.domain");
+		String configRouteDomain = XTFConfig.get("xtf.config.route_domain");
+		String systemEnvDomain = System.getenv().get("DOMAIN");
+
+		return configDomain != null ? configDomain : (systemEnvDomain != null ? systemEnvDomain : configRouteDomain);
```
Comment (maschmid) [inline core/src/main/java/cz/xtf/core/config/BackwardCompatibility.java]: The backwards compatibility interpretation of DOMAIN (or xtf.config.domain) should be

routeDomain = "apps".$DOMAIN

see https://github.com/xtf-cz/xtf/blob/master/utilities/src/main/java/cz/xtf/XTFConfiguration.java#L173

Comment (iankko) [reply core/src/main/java/cz/xtf/core/config/BackwardCompatibility.java]: @maschmid Right, great catch! Thanks, Marek. Should be fixed now.

File: core/src/main/java/cz/xtf/core/openshift/OpenShift.java [hunk 4]
```diff
@@ -393,7 +394,9 @@ public boolean deleteRoute(Route route) {
 	 * @return Hostname as is expected to be generated by OpenShift
 	 */
 	public String generateHostname(String routeName) {
-		if(routeSuffix == null) {
+		if (StringUtils.isNotBlank(cz.xtf.core.config.OpenShiftConfig.routeDomain())) {
+			routeSuffix = cz.xtf.core.config.OpenShiftConfig.routeDomain();
+		} else if (routeSuffix == null) {
```
Comment (Crumby) [inline core/src/main/java/cz/xtf/core/openshift/OpenShift.java]: Could you please move this check out off this `if() {} else if () {}` block? This way you are checking every time you call the method and do the reassignment, but it's rather lazy initialization and any subsequent call will get you same result.

**Pull request merged**